File: leetcode2/CorrectaBinaryTree/a.py
# -*- coding: utf-8 -*-
# Copyright (C) 2020 by iamslash 

# 188ms 46.74% 27.4MB 38.04%
# BFS
# O(N) O(N)
class Solution:
    def correctBinaryTree(self, root: TreeNode) -> TreeNode:
        q = [root]
        while q:
            child_nodes = []
            for u in q:
                if u.left:
                    child_nodes.append(u.left)
                if u.right:
                    child_nodes.append(u.right)
            for u in child_nodes:
                if u.right in child_nodes:
                    for p in q:
                        if p.left == u:
                            p.left = None
                        if p.right == u:
                            p.right = None
                    return root
            q = child_nodes
            
# 184ms 58.70% 27.6MB 38.04%
# BFS
# O(N) O(N)
class Solution:
    def correctBinaryTree(self, root: TreeNode) -> TreeNode:
        q = deque([root])
        while q:
            u = q.popleft()
            if u.right:
                if u.right.right in q:
                    u.right = None
                    return root
                else:
                    q.append(u.right)
            if u.left:
                if u.left.right in q:
                    u.left = None
                    return root
                else:
                    q.append(u.left)
        return root

File: leetcode2/CorrectaBinaryTree/a.cpp
/* Copyright (C) 2020 by iamslash */

#include <cstdio>

//             8
//          /     \
//        3        1
//       /        / \
//      7        9   4
//     /            / \
//    2            5   6
//
// ust: 4 5 6

// 196ms 47.54% 86MB 34.43%
// DFS
// O(N) O(N)
class Solution {
 private:
  unordered_set<TreeNode*> ust;
 public:
  TreeNode* correctBinaryTree(TreeNode* u) {
    // base
    if (!u) {
      return NULL;
    }
    if (u->right != NULL && ust.count(u->right)) {
      return NULL;
    }
    // recursion
    u->right = correctBinaryTree(u->right);
    u->left = correctBinaryTree(u->left);
    ust.insert(u);
    return u;
  }
};

// 180ms 73.77% 75.4MB 83.61%
// BFS
// O(N) O(N)
class Solution {
 public:
  TreeNode* correctBinaryTree(TreeNode* root) {
    unordered_set<TreeNode*> seen;
    queue<TreeNode**> q;
    if (root)
      q.push(&root);
    while (q.size()) {
      auto u = q.front(); q.pop();
      if ((*u)->right) {
        if (seen.count((*u)->right)) {
          *u = NULL;
          break;
        }
        seen.insert((*u)->right);
        q.push(&(*u)->right);
      }
      if ((*u)->left != NULL) {
        seen.insert((*u)->left);
        q.push(&(*u)->left);
      }
    }
    return root;
  }  
};
